### ISSUE TITLE
Crashes in combat consistently with sky reclamation project installed. (specifically the scripts)

### ISSUE THREAD

**Issue by LethalManBoob** [OpenXRay/xray-16#768]
**Describe the bug**
Crashes in combat consistently with sky reclamation project installed. (specifically the scripts)

**To Reproduce**
1. Start shooting bandits
2. Kill 2 bandits
3. Crash on the 3rd or 4th.

**Expected behavior**
Black screen. 

**BugTrap error report**
xrcore.dll is pointed to as the cause
breakpoint was discovered

**Desktop (please complete the following information):**
 - OS: Windows 10 latest possible 20h1
 - OpenXRay build version latest stable public release 64x

**Comment by LethalManBoob** [comment]
I specifically use the latest binary's from the github. Not the old version from moddb. It works fine without the openxray engine.

**Comment by Xottab-DUTY** [comment]
Any crash reports?

**Comment by LethalManBoob** [comment]
> Any crash reports?

Turns out it also crashed a LOT when you first climb the watchtower. This is not the error im getting but a different additional one. Ill get the other one if i can get past this error. Also the zip crash folder isnt accessible for me. 

SCRIPT RUNTIME ERROR
stack traceback:
	...n\stalker clear sky\gamedata\scripts\task_objects.script:1706: in function 'check_task'
	...n\stalker clear sky\gamedata\scripts\task_manager.script:356: in function <...n\stalker clear sky\gamedata\scripts\task_manager.script:348>
 
FATAL ERROR
 
[error] Expression    : !m_error_code
[error] Function      : raii_guard::~raii_guard
[error] File          : C:\projects\xray-16\src\xrScriptEngine\script_engine.cpp
[error] Line          : 569
[error] Description   : ...n\stalker clear sky\gamedata\scripts\task_objects.script:1706: attempt to index upvalue 'sim' (a nil value)

So i began deleting every script that showed up on the error report. im sure at least 90% of them have an error as i went through deleting 50% of the scripts until it told me this and i gave up.

FATAL ERROR
 
[error] Expression    : <no expression>
[error] Function      : CInifile::r_section
[error] File          : C:\projects\xray-16\src\xrCore\xr_ini.cpp
[error] Line          : 729
[error] Description   : fatal error
[error] Arguments     : Can't open section 'ammo_sections' (only 'army' avail). Please attach [*.ini_log] file to your bug report

**Comment by LethalManBoob** [comment]
I have another question. You know how to fix this rendering error im getting?
https://i.imgur.com/vWaJweR.jpg You can see a line of hard light cascade on the ground. Only disabling the sun rays fixes this and i dont want to do that.

**Comment by Decane** [comment]
>attempt to index upvalue 'sim' (a nil value)

In the SRP, task_objects.script sets `sim = _G.sim` at file-scope as soon as task_objects.script is loaded into memory. This loading-into-memory happens when CRandomTask:__init() in task_manager.script tries to call any of several possible functions in task_objects.script.

_G.start_game_callback() sets `_G.sim = alife()` when the game loads. With the vanilla engine, this always happens before CRandomTask:__init() gets called, and hence `sim` always ends up being valid in task_objects.script.

@Xottab-DUTY Can CRandomTask:__init() get called before _G.start_game_callback() during the loading sequence in OpenXRay?

@LethalManBoob If the crash is reproducible, you may be able to answer my question by inserting a debug statement like
```
error_log("start_game_callback()")
```
to the top of _G.start_game_callback(), and another one like
```
error_log("CRandomTask:__init()")
```
to the top of CRandomTask:__init(). Then, observe the order of the debug entries appearing _after_ "SERVER: Starting..." in your console log. If "CRandomTask:__init()" appears before "start_game_callback()", the load order in OpenXRay differs from vanilla.

**Comment by LethalManBoob** [comment]
Alright, so the crash i was getting was without the engine as well. Turns out it was my launch options. Ill try and reproduce the error tomorrow. With the combat and the tower, without my launch options. The whole thing has been a nightmare for me.

**Comment by Xottab-DUTY** [comment]
Still, you didn't sent the BugTrap crash reports, they may be helpful.

@LethalManBoob, what are your launch options?

**Comment by LethalManBoob** [comment]
-keep_lua -nes_texture_storing -mt_cdb -cache_cdb   
I guessed that keep Lua was for performance, texture storing for performance, mt was multithreaded cdb and the cache one was officially for performance. I removed them and all is well so far. 

like i said ill try to reproduce all bugs (with the bug reports because i deleted everything in anger)  tomorrow. 
For the record i think mt_cdb was causing the crashes. but we will know tomorrow, unless you wanna try. I just had sky reclamation project and shaders max installed when i was getting the tower crash. Not even with the openxray installed

**Comment by Xottab-DUTY** [comment]
@LethalManBoob, will be looking forward :)

So, only `-keep_lua` can possibly affect that, other options are non-relavant to scripts at all.
1. `-mt_cdb` does nothing useful – just moves CDB data generation from main thread to another thread.. while making the main thread wait until generation will be finished. (maybe, in vanilla engine it did some speedup, but not in OpenXRay)
2. It should be `-cdb_cache`, not reversed `-cache_cdb` 😛 
3. `-nes_texture_storing` may speedup game saves loading after the first one. (will not speedup the very first load)
3. `-keep_lua` – by default, engine resets all scripts and reloads them again when you load the game again, but with this key it doesn't do that.

**Comment by Decane** [comment]
>-keep_lua – by default, engine resets all scripts and reloads them again when you load the game again, but with this key it doesn't do that.

That would cause problems for sure. There are lots of scripts whose load() function e.g. doesn't explicitly empty out Lua tables set in a previous game session and instead relies on the script being reloaded automatically for the tables to be cleared. This isn't even exclusive to the SRP; the vanilla scripts also aren't safe for `-keep_lua`.

**Comment by LethalManBoob** [comment]
https://drive.google.com/file/d/1JtcL33pkGKobjULiO2_d6g6AZQam8vql/view?usp=sharing
This includes everything to do with my crash. I only installed the sky reclamation project and no launch options were used.
Everything just comes to a hard stop and no bug report is generated. I need to use the task manager to close the application.

**Comment by Decane** [comment]
In openxray_upton.log, two things jump out at me:

1.
```
! [LUA]  CharacterCommunity available only for InventoryOwner
stack traceback:
	...mmon\stalker clear sky\gamedata\scripts\xr_danger.script:72: in function <...mmon\stalker clear sky\gamedata\scripts\xr_danger.script:38>
```
That looks like a benign error raised by `danger_obj:character_community()` when `danger_obj` can't be cast to a CInventoryOwner, which I presume is true of any non-creature entity, e.g. a grenade or sound. You might be able to get rid of it by modifying [line 72 of xr_danger.script](https://github.com/Decane/SRP/blob/40df7c62514f6e6f923e87541550b71878d08d6d/gamedata/scripts/xr_danger.script#L72) to be:
```
if IsStalker(danger_obj) and danger_obj:character_community() == self.faction then
```
To be clear, this should have nothing to do with your crash; it's just a way to remove some noise from your console log.

2.
```
FATAL ERROR
 
[error] Expression    : <no expression>
[error] Function      : handler_base
[error] File          : C:\projects\xray-16\src\xrCore\xrDebug.cpp
[error] Line          : 869
[error] Description   : pure virtual function call
```
This would be the reason your game crashed. Unfortunately, "pure virtual function call" is about as unspecific as it gets. Hopefully Xottab-DUTY can help you.

**Comment by Xottab-DUTY** [comment]
> which I presume is true of any non-creature entity, e.g. a grenade or sound.

@Decane, in SOC and CS monsters had inventory, but in COP it was uprooted from the engine and we still didn't restored it. So, this log spam is most probably about mutants and some vanilla scripts that trying to access their inventory which doesn't exist...
(#677)

> This would be the reason your game crashed. Unfortunately, "pure virtual function call" is about as unspecific as it gets. Hopefully Xottab-DUTY can help you.

Yes, will look at the crash report! 🙂 

> @Xottab-DUTY Can CRandomTask:__init() get called before _G.start_game_callback() during the loading sequence in OpenXRay?

And sorry for not responding this. I forgot to answer :D
What I can notice, that there can be some ordering issues in Shadow of Chernobyl because of old LuaJIT version. (see https://github.com/OGSR/OGSR-Engine/issues/281)
Just a notice, because it can affect us at some point...

**Comment by Xottab-DUTY** [comment]
Just looked at the crash report, the engine crashed when it tried to play physics sound:

![image](https://user-images.githubusercontent.com/2989212/114216015-0046ed80-996f-11eb-9f8b-836e3aacc2a6.png)
*(file: xrGame/physics_game.cpp)*

Unfortunately, this is vanilla crash, which was hidden in the original engine.
Not intentionally, but this is how engine was configured – disabled C++ exceptions – ignore errors for maximum performance. It is a common practice in gamedev to disable C++ exceptions for release versions of the game.
But ignoring errors can lead to memory corruptions, random crashes and saves corruptions...

GSC was aware of this crash, but they didn't find the crash reason, probably.
There's no such crash in COP just because... It seems that they just stopped to use engine feature that was causing this crash.

Maybe @KRodinn or @v2v3v4 know something about this crash?

**Comment by Xottab-DUTY** [comment]
@LethalManBoob, can you test the game again with the latest Master Gold builds?

**Comment by LethalManBoob** [comment]
> @LethalManBoob, can you test the game again with the latest Master Gold builds?

I can but so far I'm only able to access the 20 December build that crashed on me. 
Unless we are due a new build very soon or you plan to send it to me.

**Comment by Xottab-DUTY** [comment]
Oh, yes... The AppVeyor was probably still building the builds when you tried to download it and you probably didn't found a link to GitHub Actions.
Just in case, both links to the AppVeyor and GitHub actions, where you can download latest "nightly" builds, are available on installation instruction page. 🙂 

The direct link to the download page from the AppVeyor: https://ci.appveyor.com/project/OpenXRay/xray-16/build/job/fp2oawx2612riugb/artifacts. The archive to download is `OpenXRay.Gold 64-bit.7z`.

**Comment by LethalManBoob** [comment]
![CRASH](https://user-images.githubusercontent.com/30273211/120889208-ec78da00-c5f3-11eb-89af-c8cae8741d3c.png)
I crash upon booting even vanilla clear sky.
[OpenXRay_error_report_210605-115345.zip](https://github.com/OpenXRay/xray-16/files/6602439/OpenXRay_error_report_210605-115345.zip)

**Comment by Xottab-DUTY** [comment]
@LethalManBoob, you didn't installed it correctly. From what engine tells: gamedata for Clear Sky is missing.

**Comment by LethalManBoob** [comment]
Turns out i have to boot without the xray engine installed to first set my settings.

**Comment by LethalManBoob** [comment]
Well, I think I had a combat crash, not 100% but I wasn't able to reproduce it, as to save games are NOT loadable after the first level transition. Upon getting to the syncing stage, the game crashes. 
[OpenXRay_error_report_210605-231843 (2).zip](https://github.com/OpenXRay/xray-16/files/6603324/OpenXRay_error_report_210605-231843.2.zip)

**Comment by Xottab-DUTY** [comment]
> Well, I think I had a combat crash, not 100% but I wasn't able to reproduce it, as to save games are NOT loadable after the first level transition. Upon getting to the syncing stage, the game crashes.
> [OpenXRay_error_report_210605-231843 (2).zip](https://github.com/OpenXRay/xray-16/files/6603324/OpenXRay_error_report_210605-231843.2.zip)

The archives inside the archive are broken :(

**Comment by LethalManBoob** [comment]
Its all I can get from the crashes. You can try it yourself. Get to the
second part where you first use your gun, or maybe when you first see the
bandits and saves WILL 100% be unusuable

**Comment by LethalManBoob** [comment]
I self built from the source code and im trying to trigger the save bug consistently on vanilla, so far, it seems to not be happening now. Very strange, ill try with the sky reclemation project later
Edit: i managed to get quite far past the first bandit fight but the save eventually became unusuable
The most recent usable crash log i have is this, but im not sure if it has anything to do with the save crash.
[OpenXRay_error_report_210608-125400.zip](https://github.com/OpenXRay/xray-16/files/6616356/OpenXRay_error_report_210608-125400.zip)

**Comment by Xottab-DUTY** [comment]
That broken saves issue may be related to table sorting order in LuaJIT. (https://github.com/OGSR/OGSR-Engine/issues/281 and https://github.com/LuaJIT/LuaJIT/issues/719)

**Comment by GiulianoB-1** [comment]
I ran into the same problem that my game crush when I use SRP. For the same cause, fighting some enemies just cause the crash. Those are 3 examples for crashes I got for the same reason(I believe), I am uploading 3 zip files just in case.
[OpenXRay_error_report_210627-223531.zip](https://github.com/OpenXRay/xray-16/files/6722229/OpenXRay_error_report_210627-223531.zip)
[OpenXRay_error_report_210627-221010.zip](https://github.com/OpenXRay/xray-16/files/6722230/OpenXRay_error_report_210627-221010.zip)
[OpenXRay_error_report_210627-221016.zip](https://github.com/OpenXRay/xray-16/files/6722234/OpenXRay_error_report_210627-221016.zip)

**Comment by GiulianoB-1** [comment]
Well, It seems like I found a semi-fix for this issue with SRP. I searched in the discord server for the word SRP to see if anyone discussed about similar problem, and I saw a message from months ago from @Xottab-DUTY  about using SRP-Lite(the message had nothing to do with this issue) instead of the normal SRP because it wasn't tested(merging openxray and srp). So I tried it, unfortunately I had to start a new save but luckily it did fix the issue. Nevertheless, it doesn't mean this issue shouldn't be fixed, SRP is still the full version, and should be used instead of SRP-Lite, but it can be used till it will be fixed. Unfortunately, I'm afraid that SRP-Lite doesn't cover all the bugs that SRP does, which is another reason why this issue should be fixed.

**Comment by Decane** [comment]
> [OpenXRay_error_report_210627-221016.zip](https://github.com/OpenXRay/xray-16/files/6722234/OpenXRay_error_report_210627-221016.zip)

The archive above is the only one to contain OpenXRay_Giuliano.log. That contains:
```
! [LUA][ERROR]  ERROR: INCORRECT LOAD [_CGeneralTask].[212][17]
stack traceback:
	...ames\s.t.a.l.k.e.r. clear sky\gamedata\scripts\_g.script:974: in function 'abort'
	...ames\s.t.a.l.k.e.r. clear sky\gamedata\scripts\_g.script:1244: in function 'set_save_marker'
	....l.k.e.r. clear sky\gamedata\scripts\task_objects.script:430: in function 'load'
	....l.k.e.r. clear sky\gamedata\scripts\task_objects.script:532: in function 'load'
	....l.k.e.r. clear sky\gamedata\scripts\task_manager.script:226: in function 'load'
	....t.a.l.k.e.r. clear sky\gamedata\scripts\se_actor.script:40: in function <....t.a.l.k.e.r. clear sky\gamedata\scripts\se_actor.script:26>
	[C]: at 0x7fffc450f8e0
	[C]: in function 'CSavedGameWrapper'
	....k.e.r. clear sky\gamedata\scripts\ui_load_dialog.script:174: in function 'file_data'
	....k.e.r. clear sky\gamedata\scripts\ui_load_dialog.script:222: in function <....k.e.r. clear sky\gamedata\scripts\ui_load_dialog.script:203>
```
@GiulianoB-1 The line numbers above imply that the **SRP isn't installed** - they are vanilla line numbers. Are you sure you have the SRP installed? Check the bottom left corner of the main menu - if it doesn't say `v1.5.10 - SRP v1.1.3`, then the SRP is not (fully) installed.

OTOH, that error can only happen on loading a save, not during combat, so perhaps you got it when attempting to load a vanilla save with the SRP installed or vice-versa.

**Comment by Xottab-DUTY** [comment]
ERROR INCORRECT LOAD happens because we have updated LuaJIT and the update contains changed behaviour...
`pairs()` now gives non-deterministic results (the iteration order is different each time), while game scripts depend on deterministic, same results...

**Comment by GiulianoB-1** [comment]
> > [OpenXRay_error_report_210627-221016.zip](https://github.com/OpenXRay/xray-16/files/6722234/OpenXRay_error_report_210627-221016.zip)
> 
> The archive above is the only one to contain OpenXRay_Giuliano.log. That contains:
> 
> ```
> ! [LUA][ERROR]  ERROR: INCORRECT LOAD [_CGeneralTask].[212][17]
> stack traceback:
> 	...ames\s.t.a.l.k.e.r. clear sky\gamedata\scripts\_g.script:974: in function 'abort'
> 	...ames\s.t.a.l.k.e.r. clear sky\gamedata\scripts\_g.script:1244: in function 'set_save_marker'
> 	....l.k.e.r. clear sky\gamedata\scripts\task_objects.script:430: in function 'load'
> 	....l.k.e.r. clear sky\gamedata\scripts\task_objects.script:532: in function 'load'
> 	....l.k.e.r. clear sky\gamedata\scripts\task_manager.script:226: in function 'load'
> 	....t.a.l.k.e.r. clear sky\gamedata\scripts\se_actor.script:40: in function <....t.a.l.k.e.r. clear sky\gamedata\scripts\se_actor.script:26>
> 	[C]: at 0x7fffc450f8e0
> 	[C]: in function 'CSavedGameWrapper'
> 	....k.e.r. clear sky\gamedata\scripts\ui_load_dialog.script:174: in function 'file_data'
> 	....k.e.r. clear sky\gamedata\scripts\ui_load_dialog.script:222: in function <....k.e.r. clear sky\gamedata\scripts\ui_load_dialog.script:203>
> ```
> 
> @GiulianoB-1 The line numbers above imply that the **SRP isn't installed** - they are vanilla line numbers. Are you sure you have the SRP installed? Check the bottom left corner of the main menu - if it doesn't say `v1.5.10 - SRP v1.1.3`, then the SRP is not (fully) installed.
> 
> OTOH, that error can only happen on loading a save, not during combat, so perhaps you got it when attempting to load a vanilla save with the SRP installed or vice-versa.

Indeed, I am using SRP-Lite. SRP is not useable with OpenXRay, tho SRP-Lite seems to be compatible. Just to be sure, I checked now if it is installed or not, and yes it is. Also, the "combat bug" didn't appear while using SRP-Lite, so for sure it is a different issue. Tho I have no idea what cause it, when those corruptions first started, I was able to "sneak by" it. The more I progressed, the more corruptions occurred, eventually it was unavoidable after I reached from the aprogrom(a bit before I was about to enter yantar, the corruptions started) to red forest. Now I just can't progress anymore :(

**Comment by Decane** [comment]
>`pairs()` now gives non-deterministic results

So far, I have found only one place in the scripts modified by the SRP where a deterministic `pairs()` iteration order is relied upon: in the save/load functions of treasure_manager.script. (The vanilla script also relies on this.)

I haven't checked if there are any vanilla scripts _not_ modified by the SRP that also rely on a deterministic iteration order.

**Comment by Xottab-DUTY** [comment]
> > `pairs()` now gives non-deterministic results
> 
> So far, I have found only one place in the scripts modified by the SRP where a deterministic `pairs()` iteration order is relied upon: in the save/load functions of treasure_manager.script. (The vanilla script also relies on this.)
> 
> I haven't checked if there are any vanilla scripts _not_ modified by the SRP that also rely on a deterministic iteration order.

We have returned old behaviour to our LuaJIT. Now we can return to the original issue with combat crashes.

@LethalManBoob, can you test the game again with the latest Master Gold build? ([direct link to OpenXRay.Gold 64-bit.7z](https://ci.appveyor.com/api/buildjobs/79awjk4cua3ay8j3/artifacts/res%2FOpenXRay.Gold%2064-bit.7z))

**Comment by LethalManBoob** [comment]
ill do this tomorrow.

**Comment by LethalManBoob** [comment]
[openxray.txt](https://github.com/OpenXRay/xray-16/files/6881021/openxray.txt)
Ok so yes it still just hard crashes mid combat. No error logs. Just a complete stop.
Also notice the major error spam in the log.

**Comment by Xottab-DUTY** [comment]
@LethalManBoob, thank you. Will try to research it more.

And the spam in the log is not related to this issue. :)

**Comment by Xottab-DUTY** [comment]
Marked this issue as critical, because the same crash appears in vanilla SOC with OpenXRay.

**Comment by globekat** [comment]
Getting back into trying OpenXray for CS, decided to install SRP to fix vanilla bugs so anything I encountered was. I am getting crashes in combat as described above (usually on 3rd or 4th NPC kill), attached latest reports for review. There hasn't been a stacktrace message in the openxray_user.log file when it crashes (sometimes the log ends midword).

I can confirm that SRP 1.1.3 has been installed (shows in main menu), using the following optional addons

- Feasable Stealth
- No enemy map spots
- Restored Pistol Ironsights
- Alternative Ballistics
- No NPC weapon tossing on death
- Real Weapon Names
- Reasonable machine gun nest difficulty
- Reduced grenade spam

I do have a few other small tweaks, mainly stuff from my own previous CS mod making that I want to try out in OpenXray since its supports, and its script bindings expose both CS and CoP functionality (hopefully). I haven't had any crashes from these yet, only when in combat.

One thing I have noticed is that when OpenXray crashes whatever image frame was currently onscreen at the time of the crash gets stuck, and I'm unable to see anything else on the desktop. This does make trying to close the bugtrap client (and by extension the game) rather difficult as I need to click in the general area of the close button. Sometimes the bugtrap client fails to appear, and I have to log off my account to clear the stuck image from the desktop.
Hopefully a smoother CTD can be implemented as well.

Using the December 2021 RC1 Modmakers release, 64bit. Have tried the players (Gold) release as well.
Using Steam version of Clear Sky
Have Windows 10 64 bit

[OpenXRay_error_report_220123-145413.zip](https://github.com/OpenXRay/xray-16/files/7919612/OpenXRay_error_report_220123-145413.zip)
[OpenXRay_error_report_220123-145403.zip](https://github.com/OpenXRay/xray-16/files/7919613/OpenXRay_error_report_220123-145403.zip)

**Comment by Decane** [comment]
^ The stack trace in the first archive's log suggests the freezing is caused by https://github.com/OpenXRay/xray-16/blob/e288845b0dba0987e4b8dd2a4f17e436ceaf9827/src/xrPhysics/MovementBoxDynamicActivate.cpp#L426 ... which only calls https://github.com/OpenXRay/xray-16/blob/311ddc7790cfc5830726bd5b6cc2e58ab0571aeb/src/xrPhysics/PHWorld.cpp#L542

Since errorlog.log cites an ACCESS_VIOLATION, I'm guessing iterator position `i` in `m_update_objects.erase(i)` is invalid. @Xottab-DUTY Am I on the right track?

As for why the most recent entry in the stack trace points at CPHUpdateObject::Deactivate() when the actual error seems to be happening in CPHWorld::RemoveUpdateObject(), I'm guessing the compiler has inlined the latter function into the former?

**Update**: I think I see why the index in the vector erase call would be invalid: In ActivateBoxDynamic(), the `vl.Deactivate()` call is sandwiched between `ph_world->Freeze()` and `ph_world->UnFreeze()` calls. But Freeze() empties m_update_objects: https://github.com/OpenXRay/xray-16/blob/311ddc7790cfc5830726bd5b6cc2e58ab0571aeb/src/xrPhysics/PHWorld.cpp#L556 ... which explains why attempting to erase an element from m_update_objects before the UnFreeze() would blow things up.

TL;DR: I think these calls in MovementBoxDynamicActivate.cpp might be the wrong way around:
```
    vl.Deactivate();
    ph_world->UnFreeze();
```

**Comment by globekat** [comment]
Another crash, this time it happened as I pressed 'F' to loot a body. No report zip, the game just froze up and the bugtrap client didn't capture it. The openxray_user.log did manage to capture it however. Looks to be a very similar stacktrace as in the report zips.

D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrPhysics.dll at 00007FF84499066E CPHUpdateObject::Deactivate() + 39998 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrPhysics.dll at 00007FF844995991 CPHUpdateObject::Deactivate() + 61281 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrPhysics.dll at 00007FF844972A63 ActivateBoxDynamic() + 16035 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrPhysics.dll at 00007FF844999B28 destroy_object_space() + 6744 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrPhysics.dll at 00007FF844999D90 destroy_object_space() + 7360 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrPhysics.dll at 00007FF84498642B RegisterGroup() + 66203 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrPhysics.dll at 00007FF844986274 RegisterGroup() + 65764 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrPhysics.dll at 00007FF84499692E CPHUpdateObject::Deactivate() + 65278 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrPhysics.dll at 00007FF8449989A9 destroy_object_space() + 2265 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrPhysics.dll at 00007FF8449986EB destroy_object_space() + 1563 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrEngine.dll at 00007FF87FA51982 CLoadScreenRenderer::OnRender() + 322 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrCore.dll at 00007FF83FF7B575 TaskManager::TaskWorkerStart() + 693 byte(s)
D:\Steam\steamapps\common\STALKER Clear Sky\bin\xrCore.dll at 00007FF83FF7C2A4 Threading::SetCurrentThreadName() + 84 byte(s)
C:\WINDOWS\System32\ucrtbase.dll at 00007FF8A4961BB2 configthreadlocale() + 146 byte(s)
C:\WINDOWS\System32\KERNEL32.DLL at 00007FF8A4FF7034 BaseThreadInitThunk() + 20 byte(s)
C:\WINDOWS\SYSTEM32\ntdll.dll at 00007FF8A6FC2651 RtlUserThreadStart() + 33 byte(s)

**Comment by globekat** [comment]
I've also been able to reproduce this error twice without the SRP, although both times the game crashed without triggering bugtrap. Once was when looting, the other on killing a Renegade. I will note that in both instances I was able to kill more renegades than I was able to with the SRP, but I'll want to test a bit more to confirm whether or not its something in the SRP that is making this error more likely to occur. It doesn't seem to occur when killing mutants.
[openxray_alex.log](https://github.com/OpenXRay/xray-16/files/7924542/openxray_alex.log)

**Comment by Xottab-DUTY** [comment]
Can't add anything detailed, but just something to add:
This issue now has a few different stack traces, related to physics and:
1. First crash in `play_object` function is a vanilla (SOC/CS/COP*) crash. (but in vanilla it was hidden, the error was there, but it did not crash, leaving space for memory and save corruptions)
It may be related to multithreading. (vanilla multithreading)

2. The newer crash logs are different, and they probably related to our changes in multithreading... And they highlighted that physics code needs refactoring.
![](https://cdn.discordapp.com/emojis/696640255679266857.webp?size=48&quality=lossless)

\* - in Call of Pripyat GSC just stopped to use the physics feature which produced the crash, but the engine still has a bug.

**Comment by Decane** [comment]
I believe I have identified and fixed the SRP regression that caused OpenXRay to crash in combat: https://github.com/Decane/SRP/commit/0d250a4c2d417e824032775d1459b046dd7bb1c7. I played the combat-heavy swamp faction war for about an hour with the following version of OpenXRay before and after the aforementioned fix:
```
OpenXRay Release Master Gold build 8605, Aug 29 2022 (Windows 64-bit, shared)
AppVeyor build 1565 (44596716) from commit[3545b5920b312c56132d913f019f22011cba420a] branch[HEAD] (built by OpenXRay)
```
Several crashes before; none after.

The crashes were apparently caused by the SRP's death_manager.script having been modified to delete an NPC's grenade(s) upon death. This was a mistake from a gameplay perspective, as the only way an NPC can ever drop a grenade in vanilla CS is if one is in slot 3 at the time of death. (Randomized grenade drops have a spawn probability of 0 in death_items_by_communities.ltx.)

From a stability standpoint, the vanilla engine tolerated this; OpenXRay does not. I believe, however, that even in OpenXRay, this regression only crashed the game when the deleted grenade was _equipped_ by the NPC upon their demise, as this would be the only time the engine and death_manager.script would both attempt to delete the same grenade. In the case of an _unequipped_ grenade in slot 3, the engine would not try to delete it; only death_manager.script would.

To anyone attempting to play CS with the SRP + OpenXray, I recommend using the [head of the SRP v1.1.4 master branch](https://github.com/Decane/SRP/archive/refs/heads/master.zip). SRP v1.1.3 will crash in combat, as will every version between v1.0.2 and v1.1.3 (inclusive) when the optional 'No NPC Weapon Tossing Upon Death' feature is used - which is where the regression was first introduced.

**Comment by Xottab-DUTY** [comment]
> I believe I have identified and fixed the SRP regression that caused OpenXRay to crash in combat: [Decane/SRP@0d250a4](https://github.com/Decane/SRP/commit/0d250a4c2d417e824032775d1459b046dd7bb1c7). I played the combat-heavy swamp faction war for about an hour with the following version of OpenXRay before and after the aforementioned fix:
> 
> ```
> OpenXRay Release Master Gold build 8605, Aug 29 2022 (Windows 64-bit, shared)
> AppVeyor build 1565 (44596716) from commit[3545b5920b312c56132d913f019f22011cba420a] branch[HEAD] (built by OpenXRay)
> ```
> 
> Several crashes before; none after.
> 
> The crashes were apparently caused by the SRP's death_manager.script having been modified to delete an NPC's grenade(s) upon death. This was a mistake from a gameplay perspective, as the only way an NPC can ever drop a grenade in vanilla CS is if one is in slot 3 at the time of death. (Randomized grenade drops have a spawn probability of 0 in death_items_by_communities.ltx.)
> 
> From a stability standpoint, the vanilla engine tolerated this; OpenXRay does not. I believe, however, that even in OpenXRay, this regression only crashed the game when the deleted grenade was _equipped_ by the NPC upon their demise, as this would be the only time the engine and death_manager.script would both attempt to delete the same grenade. In the case of an _unequipped_ grenade in slot 3, the engine would not try to delete it; only death_manager.script would.
> 
> To anyone attempting to play CS with the SRP + OpenXray, I recommend using the [head of the SRP v1.1.4 master branch](https://github.com/Decane/SRP/archive/refs/heads/master.zip). SRP v1.1.3 will crash in combat, as will every version between v1.0.2 and v1.1.3 (inclusive) when the optional 'No NPC Weapon Tossing Upon Death' feature is used - which is where the regression was first introduced.

Thanks for the info and the research!
It can help with the bug fix in the engine...

So, starting from X-Ray 1.5, physics are being processed in the secondary thread, in parallel with the main thread.
Briefly, it looks like something like this happening:
1. Main thread adds a grenade to physics processor
2. Script deletes the grenade
3. Secondary thread tries to access a pointer to already deleted object.

It's really worth to mention again, that original engine is tolerating access to a null pointer, reads and writes to the address of already deleted object... and this can lead to anything: from literally nothing to a memory corruption, which can lead to a save corruption 😥

**Comment by GiulianoB-1** [comment]
> I believe I have identified and fixed the SRP regression that caused OpenXRay to crash in combat: [Decane/SRP@0d250a4](https://github.com/Decane/SRP/commit/0d250a4c2d417e824032775d1459b046dd7bb1c7). I played the combat-heavy swamp faction war for about an hour with the following version of OpenXRay before and after the aforementioned fix:
> 
> ```
> OpenXRay Release Master Gold build 8605, Aug 29 2022 (Windows 64-bit, shared)
> AppVeyor build 1565 (44596716) from commit[3545b5920b312c56132d913f019f22011cba420a] branch[HEAD] (built by OpenXRay)
> ```
> 
> Several crashes before; none after.
> 
> The crashes were apparently caused by the SRP's death_manager.script having been modified to delete an NPC's grenade(s) upon death. This was a mistake from a gameplay perspective, as the only way an NPC can ever drop a grenade in vanilla CS is if one is in slot 3 at the time of death. (Randomized grenade drops have a spawn probability of 0 in death_items_by_communities.ltx.)
> 
> From a stability standpoint, the vanilla engine tolerated this; OpenXRay does not. I believe, however, that even in OpenXRay, this regression only crashed the game when the deleted grenade was _equipped_ by the NPC upon their demise, as this would be the only time the engine and death_manager.script would both attempt to delete the same grenade. In the case of an _unequipped_ grenade in slot 3, the engine would not try to delete it; only death_manager.script would.
> 
> To anyone attempting to play CS with the SRP + OpenXray, I recommend using the [head of the SRP v1.1.4 master branch](https://github.com/Decane/SRP/archive/refs/heads/master.zip). SRP v1.1.3 will crash in combat, as will every version between v1.0.2 and v1.1.3 (inclusive) when the optional 'No NPC Weapon Tossing Upon Death' feature is used - which is where the regression was first introduced.

I tried your suggestion and used srp 1.1.4 and it actually works just fine. no crash at combat occurs at all, finally. I did notice a different problem tho but it is more manageable. Quick saves are not usable, just clicking on a quick save makes the game crash(manual saving works just fine). I downloaded a mod that allows to do quick saves by clicking on 's' when on the menu, those saves are corruption-free.
